cmd/fanoutsidecar/main.go

```diff
@@ -125,57 +126,73 @@ func main() {
 	}
 }
 
-func setupConfigMapNoticer(logger *zap.Logger, configUpdated swappable.UpdateConfig) (manager.Manager, error) {
+func setupChannelWatcher(logger *zap.Logger, configUpdated swappable.UpdateConfig) (manager.Manager, error) {
 	mgr, err := manager.New(config.GetConfigOrDie(), manager.Options{})
 	if err != nil {
-		logger.Error("Error starting manager.", zap.Error(err))
-		return nil, err
-	}
-
-	switch configMapNoticer {
-	case cmnfVolume:
-		err = setupConfigMapVolume(logger, mgr, configUpdated)
-	case cmnfWatcher:
-		err = setupConfigMapWatcher(logger, mgr, configUpdated)
-	default:
-		err = fmt.Errorf("need to provide the --config_map_noticer flag (valid values are %s)", configMapNoticerValues())
-	}
-	if err != nil {
+		logger.Error("Error creating new maanger.", zap.Error(err))
 		return nil, err
 	}
+	v1alpha1.AddToScheme(mgr.GetScheme())
+	channelwatcher.New(mgr, logger, updateChannelConfig(configUpdated))
 
 	return mgr, nil
 }
 
-func setupConfigMapVolume(logger *zap.Logger, mgr manager.Manager, configUpdated swappable.UpdateConfig) error {
-	cmn, err := filesystem.NewConfigMapWatcher(logger, filesystem.ConfigDir, configUpdated)
-	if err != nil {
-		logger.Error("Unable to create filesystem.ConifgMapWatcher", zap.Error(err))
-		return err
+func updateChannelConfig(updateConfig swappable.UpdateConfig) channelwatcher.WatchHandlerFunc {
+	return func(ctx context.Context, c client.Client, chanNamespacedName types.NamespacedName) error {
+		channels, err := listAllChannels(ctx, c)
+		if err != nil {
+			logging.FromContext(ctx).Info("Unable to list channels", zap.Error(err))
+			return err
+		}
+		config := multiChannelFanoutConfig(channels)
+		return updateConfig(config)
 	}
-	if err = mgr.Add(cmn); err != nil {
-		logger.Error("Unable to add the config map watcher", zap.Error(err))
-		return err
-	}
-	return nil
 }
 
-func setupConfigMapWatcher(logger *zap.Logger, mgr manager.Manager, configUpdated swappable.UpdateConfig) error {
-	kc, err := kubernetes.NewForConfig(mgr.GetConfig())
-	if err != nil {
-		return err
+func listAllChannels(ctx context.Context, c client.Client) ([]v1alpha1.Channel, error) {
+	channels := make([]v1alpha1.Channel, 0)
+	cl := &v1alpha1.ChannelList{}
+	if err := c.List(ctx, &client.ListOptions{}, cl); err != nil {
+		return nil, err
+	}
+	for _, c := range cl.Items {
+		if c.Status.IsReady() && shouldWatch(&c) {
+			channels = append(channels, c)
+		}
```

Comment:
Does that work with controller runtime when channels are read from in-memory cache?

Reply:
I don't know if controller runtime actually uses the continue field today, but I think the contract it shows via the types it consumes and returns imply that it might be required. To be safe, I would prefer to follow the contract, even if in the current implementation it never gets called.